Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +17,66 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return currentAmpHours;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
-{
-    return false;
+{   
+    if( current < 0 )
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double initialValueOfCharge = initialStateOfChargePercent_ * BATTERY_AMP_HOUR_CAPACITY; //initial state in miliseconds
+    double averageIncreaseValue = totalAmpHoursUsed() * 3.6e+6 ; //the increase in charge
+    if(averageIncreaseValue <= 0)
+    {
+        averageIncreaseValue = -averageIncreaseValue;
+    }
+    if( isCharging() )
+    {
+        // Ah / a
+        double batteryChargeRemaining = (averageIncreaseValue * 3.6e+6) * (BATTERY_AMP_HOUR_CAPACITY-initialValueOfCharge) / -current; //BatteryAMPHOUR should be totalAMpHour
+        double timeLeft = batteryChargeRemaining / averageIncreaseValue;
+        QTime timeUntilFinished = currentTime.addMSecs(timeLeft);
+        return timeUntilFinished;
+    }
+    else
+    {
+        //Ah x V / Watts
+        double timeLeft = ((averageIncreaseValue * voltage * 3.6e+6 * initialValueOfCharge) / -(voltage * current));    //AH value passed around
+        QTime timeUntilFinished = currentTime.addMSecs(timeLeft);
+        return timeUntilFinished;
+    }
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    {
+        voltage = batteryData.voltage;
+        current = batteryData.current; //get the current value
+        if(currentTime.toString() == NULL)
+        {
+            previousTime = QTime(0,0,0);
+        }
+        else
+        {
+            previousTime = currentTime;
+        }
+        currentTime = batteryData.time;
+        timeDifferenceStamp = currentTime.toString( "hhmmss.zzz").toDouble() - previousTime.toString( "hhmmss.zzz" ).toDouble();
+
+        currentAmpHours = currentAmpHours + current/timeDifferenceStamp;
```

Comment:
An amphour calculation can't be done with only data point. So finding the change in time is correct, but a change in current must be taken into account as well. 

Take for example these two data points:
t=0s, I=100mA
t=10s, I=0mA

By the approach given here, the calculation be we change in time (10s), and the last current (0mA). This equates to a change in 0 amp hours, but it is probably not reasonable to assume that in that whole 10s that no current was being used.